### ISSUE TITLE
[Feature] allow to NOT restart before and after hooks on test failure and rerun

### ISSUE THREAD

**Issue by MaksymChepilDevPro** [microsoft/playwright#14586]
A bit of info from the documentation:
BeforeAll hook [test.beforeAll(hookFunction)](https://playwright.dev/docs/test-retries). 
Note that worker process is restarted on test failures, and beforeAll hook runs again in the new worker. Learn more about 

Test retries [Failures](https://playwright.dev/docs/test-retries#failures)
Should any test fail, Playwright Test will discard the entire worker process along with the browser and will start a new one. Testing will continue in the new worker process starting with the next test. 
> 
> Worker process #1 starts
> beforeAll hook runs
> first good passes
> second flaky fails

> Worker process #2 starts
> beforeAll hook runs again
> third good passes

I understand the logic of the rerun all the scenarios, nevertheless in my case beforeHooks made a lot of precondition moves.

As a made-up example: 
We need to test a feature with participants in the event.
A new feature is to edit participants (remove/add).

**Preconditions:(Before Hook)**

1. We need to create a general event that can be a few days long 
2. Add a meeting to this event 
3. invite participants and confirm the invitation via email.

**Tests**

**After Hooks:**
In after hooks we are deleting:

1. all the participants
2. the meeting
3. the event

As a result, we have a URL for the created meeting where we can do our test, and we can navigate to the page via URL.
All the test are independent, nevertheless, they all in this test suite has one general precondition that creates the event, etc.

So if one of the tests fails due to the wrong title text it will trigger the afterAll hook and delete all the data, and then the next worker will proceed with starting the before hook and creating new fresh data for the test. 
**That makes sense** when your preconditions are has a low performance for the application such as simple navigation to the page without creating something.
Nevertheless, it's a pain in the back, when before hook making a lot of preconditions, and after hook cleaning it up.

**So my feature request is to allow to run the tests and do not restart hooks per each worker.** 

P.s. This logic - with NOT rerun hooks after one of the tests is failed is working by default in Jest. 
After we migrated to the playwright test runner we faced a lot of performance issues and due to the hook behavior, we faced a test run time decrease.
For now, we migrated back to the Jest and looking forward to returning to the native PW test runner.

If you have any further questions, feel free to ask them/

**Comment by aslushnikov** [comment]
cc'ing @dgozman as this might be another use-case for global fixtures.

**Comment by szamacz** [comment]
From my side, I would add that while it is a great idea to restart the worker in case of failed UI test, this doesn't make much sense when it comes to failed API test.

**Comment by pavelfeldman** [comment]
Heavy test setup should probably done outside of the test run, otherwise entire dev workflow becomes compromised. Imagine that every time user hits a "run" icon in the VS Code sidebar, against a single test, this entire setup needs to be happening. We'd like our tests to run fast and be easy to debug, and this request is conflicting with our testing vision / priorities. Hence closing as won't fix.

**Comment by MaksymChepilDevPro** [comment]
@pavelfeldman 
Ok, so you suggest creating a heavy test setup outside of the test run. 
And what if we need to make this "heavy" setup per each test suite that aims to test unique flow?

For now, we have more than 300 suites like this. 
In this case, we cannot use the `PW test runner` by default and need to use separate test runners.

`We'd like our tests to run fast and be easy to debug, and this request is conflicting with our testing vision / priorities. Hence closing as won't fix.` 

I'm not trying to force you to fix something just suggest improving the configuration of the hooks, nevertheless, your vision is like you creating something that planning to use only for your product flow and don't want to extend it to be more flexible than other runners. 

```
Playwright can either be used as a part of the Playwright Test test runner (this guide), or as a Playwright Library

Playwright Test was created specifically to accommodate the needs of the end-to-end testing.
 It does everything you would expect from the regular test runner, and more. Playwright test allows to:

Run tests across all browsers.
Execute tests in parallel.
Enjoy context isolation out of the box.
Capture videos, screenshots and other artifacts on failure.
Integrate your POMs as extensible fixtures.
```

So, actually, the PW test runner does not do everything that you would expect from the regular test runner. 
And basically aims to run simple scenarios and it is recommended to use it only for that. 

All the heavy setup for the tests used in "before and after hooks" will:
- decreasing the time of the test suite run 
- increase performance load via restarting "before and after hooks" after each failed test in this test suite - even if rerun is disabled

`Note!` Hope you will extend Playwright outside the PW test runner as well as in tight integration

**Comment by ipappous** [comment]
Hi all, 
We have a similar case. 
We are able to deploy and destroy a test db in a matter of seconds. We can use this DB for x atomic test cases. 
It would be very nice to be able to perform this setup before our test suite will run from within Playwright, instead of creating hacks with scripts. 

mk

**Comment by Gennadiii** [comment]
I assume there's no solution for this issue? I get it now why people use jest as test runner. This is really annoying the be forced to waste such a lot of time on browser relaunch when my beforeAll is god enough for the rest of the tests.

**Comment by philipfong** [comment]
It's unfortunate that there isn't much of an in-between solution of the current beforeAll/afterAll behavior versus global setup. And how bizarre for the beforeAll/afterAll behavior to deviate so far from what we've seen in Jest for so long.

**Comment by trix7xter** [comment]
If you do not test anything more complex than a lingerie store, then indeed, this modification is not needed.

Why close such a valuable request? Let's assume I have automated tests that use the same object, created in beforeAll and removed in afterAll. Why should I recreate it if a test fails? This would disrupt the execution of other tests.

The creation of this object is very resource-intensive, and it contains **unique** and **limited** properties that cannot exist in other objects. It's impossible to create such a parameterized object for every test.

**Comment by MarcRosewarne** [comment]
Although not elegant, perhaps use `testInfo.retry` to check if first run or a retry in the `beforeAll` hook.

```
test.beforeAll(async ({ browser, browserName }, testInfo) => {
  console.log(`RUNNING: ${testInfo.title} (${browserName})`);
  console.log("RETRY COUNT: " + testInfo.retry);
  if(testInfo.retry == 0){
    //do the beforeAll on first try stuff here
  }
});
```
If you need to be more specific you could use `testInfo.title` as an extra filter for target a specific test in the spec file.

**Comment by matt-kxs** [comment]
Can we please re-open this issue? As others have discussed, a lot of us have much more complicated e2e setups that require a lot of pre-conditions that are not cheap.

**Comment by frasu** [comment]
I have one test to check hundreds of stuff. But on the report, I don't see what causes the fail in first look.

So I decide to grab whole interested data in beforeAll section and then in loops I execute more gradated tests to check every assertion.

But because of reruning beforeAll section after single fail I cannot check all of my tests.

Please, reopen this issue.

**Comment by frasu** [comment]
> Although not elegant, perhaps use `testInfo.retry` to check if first run or a retry in the `beforeAll` hook.
> 
> ```
> test.beforeAll(async ({ browser, browserName }, testInfo) => {
>   console.log(`RUNNING: ${testInfo.title} (${browserName})`);
>   console.log("RETRY COUNT: " + testInfo.retry);
>   if(testInfo.retry == 0){
>     //do the beforeAll on first try stuff here
>   }
> });
> ```
> 
> If you need to be more specific you could use `testInfo.title` as an extra filter for target a specific test in the spec file.

Doesn't work for me. After test fail, the beforeAll section is executed again because retry count is 0 every time.

**Comment by matt-kxs** [comment]
> > Although not elegant, perhaps use `testInfo.retry` to check if first run or a retry in the `beforeAll` hook.
> > ```
> > test.beforeAll(async ({ browser, browserName }, testInfo) => {
> >   console.log(`RUNNING: ${testInfo.title} (${browserName})`);
> >   console.log("RETRY COUNT: " + testInfo.retry);
> >   if(testInfo.retry == 0){
> >     //do the beforeAll on first try stuff here
> >   }
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If you need to be more specific you could use `testInfo.title` as an extra filter for target a specific test in the spec file.
> 
> Doesn't work for me. After test fail, the beforeAll section is executed again because retry count is 0 every time.

It looks like they're going to address it as part of https://github.com/microsoft/playwright/issues/32803

**Comment by matt-kxs** [comment]
> I have one test to check hundreds of stuff. But on the report, I don't see what causes the fail in first look.
> 
> So I decide to grab whole interested data in beforeAll section and then in loops I execute more gradated tests to check every assertion.
> 
> But because of reruning beforeAll section after single fail I cannot check all of my tests.
> 
> Please, reopen this issue.

It looks like they're going to address it as part of https://github.com/microsoft/playwright/issues/32803
